cpp/src/lance/arrow/dataset_test.cc

```diff
@@ -0,0 +1,77 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "lance/arrow/dataset.h"
+
+#include <arrow/table.h>
+#include <fmt/format.h>
+
+#include <catch2/catch_test_macros.hpp>
+#include <memory>
+#include <string>
+
+#include "lance/arrow/file_lance.h"
+#include "lance/arrow/stl.h"
+#include "lance/testing/io.h"
+
+using lance::arrow::ToArray;
+
+std::shared_ptr<::arrow::Table> ReadTable(const std::string& uri, std::optional<int32_t> version) {
+  std::string path;
+  auto fs = ::arrow::fs::FileSystemFromUriOrPath(uri, &path).ValueOrDie();
+  auto actual_dataset = lance::arrow::LanceDataset::Make(fs, uri, version).ValueOrDie();
+  CHECK(actual_dataset != nullptr);
+  return actual_dataset->NewScan().ValueOrDie()->Finish().ValueOrDie()->ToTable().ValueOrDie();
+}
+
+TEST_CASE("Create new dataset") {
+  auto ids = ToArray({1, 2, 3, 4, 5, 6, 8}).ValueOrDie();
+  auto values = ToArray({"a", "b", "c", "d", "e", "f", "g"}).ValueOrDie();
+  auto table1 = ::arrow::Table::Make(::arrow::schema({::arrow::field("id", ::arrow::int32()),
+                                                      ::arrow::field("value", ::arrow::utf8())}),
+                                     {ids, values});
+
+  auto dataset = lance::testing::MakeDataset(table1).ValueOrDie();
+
+  auto base_uri = lance::testing::MakeTemporaryDir().ValueOrDie() + "/testdata";
+  auto format = lance::arrow::LanceFileFormat::Make();
+  ::arrow::dataset::FileSystemDatasetWriteOptions write_options;
+  std::string path;
+  auto fs = ::arrow::fs::FileSystemFromUriOrPath(base_uri, &path).ValueOrDie();
+  write_options.filesystem = fs;
+  write_options.base_dir = path;
+  write_options.file_write_options = format->DefaultWriteOptions();
+
+  auto status = lance::arrow::LanceDataset::Write(
+      write_options, dataset->NewScan().ValueOrDie()->Finish().ValueOrDie());
+  CHECK(status.ok());
+
+  ids = ToArray({100, 101, 102}).ValueOrDie();
+  values = ToArray({"aaa", "bbb", "ccc"}).ValueOrDie();
+  auto table2 = ::arrow::Table::Make(::arrow::schema({::arrow::field("id", ::arrow::int32()),
+                                                      ::arrow::field("value", ::arrow::utf8())}),
+                                     {ids, values});
+  dataset = lance::testing::MakeDataset(table2).ValueOrDie();
+  status = lance::arrow::LanceDataset::Write(
```

Comment:
so from the user's perspective, say they get a bunch of new images. Then they construct a table with the new rows but the same schema as the existing dataset. Then they'll call `pyarrow.dataset.write_dataset`? What write options do they need to pass in for append (vs say overwrite)?

Reply:
or is the idea that we don't support overwrite anymore?

Reply:
That's a good question.

https://arrow.apache.org/docs/python/generated/pyarrow.dataset.write_dataset.html 
https://github.com/apache/arrow/blob/master/cpp/src/arrow/dataset/type_fwd.h#L56-L66

Arrow does not have `Append` mode? Guess we can have `Write` and `Append` interfaces separately.

Reply:
Discussed offline, decided to pass a `WriteModel::{CREATE / APPEND / OVERWRITE}` instead.